File: client-python/src/bae_py/harness/transport.py
"""The HTTP seam. The harness talks to the server exclusively through a
:class:`Transport`; the default implementation wraps ``httpx.AsyncClient``, and
tests inject a mock through the same protocol so they run fully offline.
"""

from __future__ import annotations

import json as _json
from dataclasses import dataclass
from typing import Any, AsyncIterator, Mapping, Protocol, runtime_checkable

from opentelemetry import propagate

from ..errors import ApiError, TransportError

# The W3C Trace Context wire allowlist (telemetry contract §6): the only headers
# BAE ever injects. Anything else an ambient propagator emits — baggage most
# notably, or any header a custom/composite global propagator writes — is
# dropped, so no baggage value (token, tenant id, prompt fragment) leaks onto a
# BAE request. Matched case-insensitively.
_ALLOWED_PROPAGATION_HEADERS = frozenset({"traceparent", "tracestate"})


def _inject_trace_context(headers: dict[str, str]) -> None:
    """Inject the ambient W3C trace context into ``headers`` in place, restricted
    to the wire allowlist. A no-op propagator or an invalid (no-SDK) span context
    writes nothing, so this costs nothing when the embedding app installed no
    OTel SDK."""
    carrier: dict[str, str] = {}
    propagate.inject(carrier)
    for key, value in carrier.items():
        if key.lower() in _ALLOWED_PROPAGATION_HEADERS:
            headers[key] = value


@dataclass(slots=True)
class TransportResponse:
    """A raw HTTP response: the status code and the parsed JSON body (or ``None``)."""

    status: int
    body: Any


@runtime_checkable
class Transport(Protocol):
    """Minimal async HTTP interface the harness depends on.

    ``request`` covers the REST management routes (session open/close, events
    replay); ``stream`` drives the JSON-RPC session loop over ``…/rpc``,
    yielding one decoded JSON-RPC frame (a ``dict``) per NDJSON line.
    """

    async def request(
        self,
        method: str,
        url: str,
        *,
        headers: Mapping[str, str],
        json: Any | None = None,
    ) -> TransportResponse: ...

    def stream(
        self,
        method: str,
        url: str,
        *,
        headers: Mapping[str, str],
        json: Any | None = None,
    ) -> AsyncIterator[dict[str, Any]]: ...

    async def aclose(self) -> None: ...


class HttpxTransport:
    """Default transport, backed by a lazily-imported ``httpx.AsyncClient``."""

    def __init__(self, timeout: float = 30.0) -> None:
        try:
            import httpx
        except ImportError as exc:  # pragma: no cover - dependency is declared
            raise TransportError(
                "httpx is required for the default transport; install bae-py with its deps"
            ) from exc
        self._httpx = httpx
        self._client = httpx.AsyncClient(timeout=timeout)

    async def request(
        self,
        method: str,
        url: str,
        *,
        headers: Mapping[str, str],
        json: Any | None = None,
    ) -> TransportResponse:
        # Inject W3C traceparent/tracestate from the ambient OTel context, if
        # any (restricted to the wire allowlist — never baggage).
        outgoing = dict(headers)
        _inject_trace_context(outgoing)
        try:
            resp = await self._client.request(method, url, headers=outgoing, json=json)
        except self._httpx.HTTPError as exc:
            raise TransportError(f"request to {url} failed: {exc}") from exc
        body: Any = None
        if resp.content:
            try:
                body = resp.json()
            except ValueError:
                body = None
        return TransportResponse(status=resp.status_code, body=body)

    async def stream(
        self,
        method: str,
        url: str,
        *,
        headers: Mapping[str, str],
        json: Any | None = None,
    ) -> AsyncIterator[dict[str, Any]]:
        """POST a JSON-RPC request and yield each NDJSON frame as it arrives.

        A non-2xx status is a pre-stream RFC 7807 error (:class:`ApiError`, e.g.
        auth), raised before the first frame; the stream body itself is HTTP 200.
        """
        outgoing = dict(headers)
        _inject_trace_context(outgoing)
        try:
            async with self._client.stream(method, url, headers=outgoing, json=json) as resp:
                if not (200 <= resp.status_code < 300):
                    await resp.aread()
                    body: Any = None
                    if resp.content:
                        try:
                            body = resp.json()
                        except ValueError:
                            body = None
                    raise ApiError.from_body(resp.status_code, body)
                async for line in resp.aiter_lines():
                    line = line.strip()
                    if not line:
                        continue
                    try:
                        yield _json.loads(line)
                    except ValueError as exc:
                        raise TransportError(
                            f"malformed JSON-RPC frame from {method} {url}"
                        ) from exc
        except self._httpx.HTTPError as exc:
            raise TransportError(f"request to {url} failed: {exc}") from exc

    async def aclose(self) -> None:
        await self._client.aclose()

File: client-typescript/src/transport.ts
import { ApiError, TransportError } from "./errors.js";
import { injectAmbientContext } from "./telemetry.js";
import type { JsonRpcFrame, SessionEvent } from "./types.js";

/** A single HTTP request against the client surface. */
export interface TransportRequest {
  method: "GET" | "POST" | "DELETE";
  /** Path beginning with `/`, e.g. `/api/v1/sessions`. */
  path: string;
  /** Bearer token (client key or session key). */
  token: string;
  /** JSON request body, if any. */
  body?: unknown;
}

/** A parsed HTTP response. `body` is the decoded JSON (or `undefined`). */
export interface TransportResponse {
  status: number;
  body: unknown;
}

/**
 * The seam the harness talks through. `request` covers the REST management
 * routes (session open/close, events replay); `stream` drives the JSON-RPC
 * session loop over `…/rpc`, yielding one decoded {@link JsonRpcFrame} per
 * NDJSON line. Tests inject a mock so the whole loop runs offline.
 */
export interface Transport {
  request(req: TransportRequest): Promise<TransportResponse>;
  stream(req: TransportRequest): AsyncIterable<JsonRpcFrame>;
}

/** `fetch`-backed transport rooted at a server base URL. */
export class FetchTransport implements Transport {
  constructor(private readonly baseUrl: string) {}

  async request(req: TransportRequest): Promise<TransportResponse> {
    const headers: Record<string, string> = {
      authorization: `Bearer ${req.token}`,
    };
    if (req.body !== undefined) {
      headers["content-type"] = "application/json";
    }
    injectAmbientContext(headers);

    let res: Response;
    try {
      res = await fetch(this.baseUrl + req.path, {
        method: req.method,
        headers,
        body: req.body !== undefined ? JSON.stringify(req.body) : undefined,
      });
    } catch (cause) {
      throw new TransportError(
        `request to ${req.method} ${req.path} failed: ${
          cause instanceof Error ? cause.message : String(cause)
        }`,
      );
    }

    const text = await res.text();
    let body: unknown;
    if (text.length > 0) {
      try {
        body = JSON.parse(text);
      } catch {
        throw new TransportError(
          `non-JSON ${res.status} response from ${req.method} ${req.path}`,
        );
      }
    }
    return { status: res.status, body };
  }

  /**
   * POST a JSON-RPC request and yield each NDJSON frame as it arrives. A
   * non-2xx status is a pre-stream RFC 7807 error ({@link ApiError}, e.g. auth)
   * raised before the first frame; the body itself is HTTP 200.
   */
  stream(req: TransportRequest): AsyncIterable<JsonRpcFrame> {
    const headers: Record<string, string> = {
      authorization: `Bearer ${req.token}`,
      "content-type": "application/json",
      accept: "application/x-ndjson",
    };
    injectAmbientContext(headers);

    return this.streamWithHeaders(req, headers);
  }

  private async *streamWithHeaders(
    req: TransportRequest,
    headers: Record<string, string>,
  ): AsyncIterable<JsonRpcFrame> {
    let res: Response;
    try {
      res = await fetch(this.baseUrl + req.path, {
        method: req.method,
        headers,
        body: req.body !== undefined ? JSON.stringify(req.body) : undefined,
      });
    } catch (cause) {
      throw new TransportError(
        `request to ${req.method} ${req.path} failed: ${
          cause instanceof Error ? cause.message : String(cause)
        }`,
      );
    }

    if (!res.ok) {
      const text = await res.text();
      let body: unknown;
      if (text.length > 0) {
        try {
          body = JSON.parse(text);
        } catch {
          throw new TransportError(
            `non-JSON ${res.status} response from ${req.method} ${req.path}`,
          );
        }
      }
      throw ApiError.fromBody(res.status, body);
    }

    if (res.body === null) return;
    const reader = res.body.getReader();
    const decoder = new TextDecoder();
    let buf = "";
    for (;;) {
      const { done, value } = await reader.read();
      if (done) break;
      buf += decoder.decode(value, { stream: true });
      let nl: number;
      while ((nl = buf.indexOf("\n")) >= 0) {
        const line = buf.slice(0, nl).trim();
        buf = buf.slice(nl + 1);
        if (line.length > 0) yield parseFrame(line, req);
      }
    }
    const last = buf.trim();
    if (last.length > 0) yield parseFrame(last, req);
  }
}

function parseFrame(line: string, req: TransportRequest): JsonRpcFrame {
  try {
    return JSON.parse(line) as JsonRpcFrame;
  } catch {
    throw new TransportError(
      `malformed JSON-RPC frame from ${req.method} ${req.path}`,
    );
  }
}

/** A frame carrying an `id` is the terminal response; anything else is a notification. */
export function isTerminalFrame(frame: JsonRpcFrame): boolean {
  return frame.id !== undefined && frame.id !== null;
}

/** Decode a `session.event` notification's `params` into a {@link SessionEvent}, else null. */
export function eventFromFrame(frame: JsonRpcFrame): SessionEvent | null {
  if (frame.method !== "session.event" || frame.params === undefined) {
    return null;
  }
  return frame.params as SessionEvent;
}

/** Throw {@link ApiError} on any non-2xx status; otherwise return the body. */
export function expectOk(res: TransportResponse): unknown {
  if (res.status < 200 || res.status >= 300) {
    throw ApiError.fromBody(res.status, res.body);
  }
  return res.body;
}
